### PR TITLE
Back port the Voice instruction to UI SDK.

### LINKED ISSUE #2508: MapboxNavigation, support `fun retrieveSsmlAnnouncementInstruction(Int)`
For `UI module`(_VoiceInstructionCache_ for sample) need `Announcement Instruction` from Native.
Provide `Ssml Announcement Instruction` from `Navigator` through `MapboxNavigation`

### PR DESCRIPTION
**Pull request by JunDai**
## Description

Backport the Voice instruction to UI SDK. - https://github.com/mapbox/navigation-sdks/issues/260
The fundamental of voice instruction implementation were already in the compatible UI SDK. This PR ports some missing methods back into `MapboxNavigation` and `NavigationViewModel`. Once the consumer calls `NavigationView#startNavigation()`, the voice instruction should work.

Also add sample code into the `SimpleMapboxNavigationKt` example to show how to use the voice instruction component separately.

closes #2508

- [x] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 1]
```diff
@@ -325,6 +325,11 @@ constructor(
         MapboxNativeNavigatorImpl.addHistoryEvent(eventType, eventJsonProperties)
     }
 
+    /**
+     *
+     */
+    fun retrieveSsmlAnnouncementInstruction(index: Int) = MapboxNativeNavigatorImpl.getVoiceInstruction(index)?.ssmlAnnouncement
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]:
```suggestion
    fun retrieveSsmlAnnouncementInstruction(index: Int): String = MapboxNativeNavigatorImpl.getVoiceInstruction(index)?.ssmlAnnouncement
```

**Comment by RingerJK** [comment]
closes https://github.com/mapbox/mapbox-navigation-android/issues/2508

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java [hunk 2]
```diff
@@ -12,75 +12,73 @@
 
 class VoiceInstructionCache {
```
Comment (RingerJK) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java]: hard to get what were changed. Could you revert to previous formatting?
Moreover, `CI` failed https://app.circleci.com/jobs/github/mapbox/mapbox-navigation-android/17755 on java-code checkstyle

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java [hunk 3]
```diff
@@ -12,75 +12,73 @@
...
-        for (VoiceInstructions ignored : step.voiceInstructions()) {
-          totalVoiceInstructions++;
+    void preCache(DirectionsRoute route) {
+        if (!connectivityStatus.isConnected()) {
+            return;
         }
-      }
-    }
-    List<String> voiceInstructionsToCache = new ArrayList<>();
-    // TODO add to MapbxoNavigation retrieveSsmlAnnouncementInstruction from Navigator
```
Comment (JunDai) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java]: @RingerJK - the change in this class is un-comment the below code and the same one below.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 1 above)
Comment (JunDai) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java (hunk 2 above)
Comment (JunDai) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java]: good catch! I think the checkstyle is different than my usual coding style regarding the indent. Just set up the Mapbox checkstyle, the code now should be clear.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java (hunk 3 above)
Comment (RingerJK) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/VoiceInstructionCache.java]: now it's obvious 🙇  thanks @JunDai !

File: examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt [hunk 4]
```diff
@@ -165,6 +177,17 @@ class SimpleMapboxNavigationKt : AppCompatActivity(), OnMapReadyCallback {
...
+        initializeSpeechPlayer()
+    }
+
+    private fun initializeSpeechPlayer() {
+        val cache = Cache(File(application.cacheDir, VOICE_INSTRUCTION_CACHE), 10 * 1024 * 1024)
+        val voiceInstructionLoader =
+            VoiceInstructionLoader(application, Mapbox.getAccessToken(), cache)
+        val speechPlayerProvider =
+            SpeechPlayerProvider(application, Locale.US.language, true, voiceInstructionLoader)
+        speechPlayer = NavigationSpeechPlayer((speechPlayerProvider))
```
Comment (RingerJK) [inline examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt]: nit 
```suggestion
        speechPlayer = NavigationSpeechPlayer(speechPlayerProvider)
```

File: examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt [hunk 5]
```diff
@@ -355,6 +380,7 @@ class SimpleMapboxNavigationKt : AppCompatActivity(), OnMapReadyCallback {
         if (mapboxNavigation.getRoutes().isEmpty() && mapboxNavigation.getTripSessionState() == TripSessionState.STARTED) {
             // use this to kill the service and hide the notification when going into the background in the Free Drive state,
             // but also ensure to restart Free Drive when coming back from background by using the channel
+            mapboxNavigation.unregisterVoiceInstructionsObserver(this)
```
Comment (RingerJK) [inline examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt]: what if put register/unregister to start/stop _trip session_ `observer`(to observe it register observer `mapboxNavigation.registerTripSessionStateObserver()`). wdyt?

Comment (JunDai) [reply examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt]: in that case, do we kind of lose the flexibility? some customers may only interest in voice instruction and some others may only care banner view update etc.?

Comment (RingerJK) [reply examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt]: IMO it looks not so obvious when register/unregister observer could register/unregister from a few places. But it's samples and I'm ok to keep it as is. up 2 you 👍

**Pull request merged**